abnf/odata-aggregation-testcases.yaml

```diff
@@ -335,6 +335,17 @@ TestCases:
     Expect:
       - aggregatableFactor:Product/Name # is also an aggrPathPrimitive
 
+  - Name: aggregate - custom aggregation method with type cast
+    Rule: queryOptions
+    Input: $apply=aggregate(Products/Self.DigitalProduct with Custom.discount as Stuff)
+    Expect:
+      - aggrPathPrefix:Products/Self.DigitalProduct # is also an aggrCollPath
+
+  - Name: aggregate - forbidden type cast after single-valued segment
+    Rule: queryOptions
+    FailAt: 44
+    Input: $apply=aggregate(Product/Self.DigitalProduct with Custom.discount as Stuff)
```

Comment:
> * `aggregate(Product/Self.DigitalProduct/$count)` counts only digital products (hence non-null).

The core spec is ambiguous here:
* `Sales('Bread')/Product/Self.DigitalProduct` evaluates to null according to [OData-CSDL, section 14.4.1.1](https://docs.oasis-open.org/odata/odata-csdl-xml/v4.01/odata-csdl-xml-v4.01.html#_Toc38530423).
* `Sales('Bread')/Product/Self.DigitalProduct` returns 404 Not Found according to [OData-URL, section 4.11](https://docs.oasis-open.org/odata/odata/v4.01/os/part2-url-conventions/odata-v4.01-os-part2-url-conventions.html#_Toc31360947).

According to the first bullet point, the null value should be counted. According to the second bullet point, it should not.

Reply:
[OData-URL, section 4.11](https://docs.oasis-open.org/odata/odata/v4.01/os/part2-url-conventions/odata-v4.01-os-part2-url-conventions.html#_Toc31360947) also states that this evaluates to null if it is (part of) a path expression, so no ambiguity regarding expression evaluation.

Returning 404 when used at the end of a resource path is somewhat odd because null-valued properties and navigation properties return 204 when used at the end of a resource path.

Reply:
> [OData-URL, section 4.11](https://docs.oasis-open.org/odata/odata/v4.01/os/part2-url-conventions/odata-v4.01-os-part2-url-conventions.html#_Toc31360947) also states that this evaluates to null if it is (part of) a path expression,

But only
> If used in a path expression **that is part of a Boolean expression**, the type cast will evaluate to null.

Reply:
I think the "that is part of a Boolean expression" is misleading/wrong and should be removed

Reply:
Opened https://issues.oasis-open.org/browse/ODATA-1563 to track this